### PR TITLE
Refact: improve task resume mechanism for graphrag

### PR DESCRIPTION
**Pull request by SHAI-minal**
### What problem does this PR solve?

Addresses review feedback on #14074 (Checkpoint mechanism for long-running workflow jobs, issue #12494).

**Changes based on @yuzhichang's review:**

1. **Renamed `checkpoint_service.py` → `task_checkpoint.py`** as suggested.
2. **Replaced Redis with direct docEngine queries** as suggested — the subgraph already gets persisted to the doc store by `generate_subgraph()`, so we just query for it instead of maintaining a separate checkpoint in Redis. This is simpler, has no extra dependency, and uses a single source of truth.

**Changes based on CodeRabbit review:**

3. **Fixed `source_id` query format mismatch** — subgraphs are stored with `source_id: [doc_id]` (list), but the original query used `source_id: doc_id` (string). Now follows the same pattern as `does_graph_contains()` in `rag/graphrag/utils.py`: filter by `knowledge_graph_kwd` only, then match `source_id` in Python. This avoids ambiguity across Elasticsearch / Infinity / OceanBase backends.

### Changes

| File | Change |
|---|---|
| `api/db/services/task_checkpoint.py` (new) | `load_subgraph_from_store()` and `has_raptor_chunks()` — docEngine-based checkpoint queries |
| `rag/graphrag/general/index.py` | `build_one()` calls `load_subgraph_from_store()` before running LLM extraction |
| `rag/svr/task_executor.py` | RAPTOR per-doc loop calls `has_raptor_chunks()` before processing |
| `test/unit_test/rag/graphrag/test_checkpoint_resume.py` (new) | 10 unit tests covering subgraph loading, source_id filtering, edge cases |

### How it works

- **GraphRAG:** Before running expensive LLM entity/relation extraction for a doc, checks the doc store for an existing subgraph (saved by a previous interrupted run). If found, loads it directly and skips LLM calls.
- **RAPTOR:** Before processing a doc, checks if RAPTOR chunks (`raptor_kwd="raptor"`) already exist for it. If yes, skips.

### Testing

- 10 new unit tests — all passing
- Full existing suite: 617 passed

### Type of change

- [x] New Feature (non-breaking change which adds functionality)
- [x] Refactoring

### CODE REVIEW

File: api/db/services/task_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+import networkx as nx
+
+from common import settings
+from common.doc_store.doc_store_base import OrderByExpr
+from common.misc_utils import thread_pool_exec
+from rag.nlp import search
+
+
+async def load_subgraph_from_store(tenant_id: str, kb_id: str, doc_id: str):
```
Comment (yuzhichang) [inline api/db/services/task_checkpoint.py]: Move functions of this file to the place where they are needed. Remove this file.

File: test/unit_test/rag/graphrag/test_checkpoint_resume.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    "rag.utils.base64_image",
+    "rag.utils.raptor_utils",
+    "rag.prompts.generator",
+    "rag.raptor",
+    "rag.app",
+    "rag.graphrag.utils",
+]
+for _m in _EXTRA_MOCKS:
+    if _m not in sys.modules:
+        sys.modules[_m] = MagicMock()
```
[the remark this answers is not in the document]
Comment (SHAI-minal) [reply test/unit_test/rag/graphrag/test_checkpoint_resume.py:80]: This follows the same sys.modules patching pattern already used in conftest.py for this test directory. Happy to revisit in a follow-up if the maintainers prefer a different approach.

**Comment by SHAI-minal** [comment]
@yuzhichang @wangq8 , Please review,

File: rag/graphrag/general/index.py [hunk 3]
```diff
@@ -43,6 +43,67 @@
 from rag.nlp import rag_tokenizer, search
 from rag.utils.redis_conn import RedisDistributedLock
 from common import settings
+from common.doc_store.doc_store_base import OrderByExpr
+
+_PAGE_SIZE = 256
+
+
+async def load_subgraph_from_store(tenant_id: str, kb_id: str, doc_id: str):
```
Comment (wangq8) [inline rag/graphrag/general/index.py:50]: @MinalMahalaShorthillsAI If we have 10,000 results and 100 `doc_id`, the current approach would run 100 queries to ES/DB, each retrieving 10,000 results, this creates significant load and inefficiency.

A better approach is to query once and process all `doc_id` in memory, reducing the number of queries and overall overhead.

An even more optimal solution is to include doc_id in the search criteria, allowing ES/DB to filter results directly. This leverages indexing and avoids fetching a large result set only to filter it afterward, significantly improving performance if search with `doc_id` supported.

File: rag/graphrag/general/index.py [hunk 4]
```diff
@@ -43,6 +43,67 @@
...
+    """Load a previously saved subgraph from the doc store.
```
Comment (yuzhichang) [inline rag/graphrag/general/index.py:51]: See rag/graphrag/general/index.py  `generate_subgraph()`
```
    chunk = {
        "content_with_weight": json.dumps(nx.node_link_data(subgraph, edges="edges"), ensure_ascii=False),
        "knowledge_graph_kwd": "subgraph",     # filter condition 1
        "kb_id": kb_id,
        "source_id": [doc_id],                                     # filter condition 2
        "available_int": 0,
        "removed_kwd": "N",
    }

```

File: rag/graphrag/general/index.py (hunk 3 above)
Comment (SHAI-minal) [reply rag/graphrag/general/index.py:50]: Thanks for the suggestion! We've now included source_id: [doc_id] directly in the search condition (as hinted by @yuzhichang 's generate_subgraph reference), so the doc store filters by index. Each call fetches at most 1 chunk with limit=1. No pagination or Python-side filtering needed.

**Comment by SHAI-minal** [comment]
@yuzhichang 
Friendly ping - happy to make any further changes if needed!

**Review by yuzhichang — APPROVED** [review]

**Review by wangq8 — APPROVED** [review]

**Comment by SHAI-minal** [comment]
@yingfeng @yuzhichang @wangq8 , It was very nice working with you all, Hoping for collaboration.
